homeassistant/components/device_tracker/legacy.py

```diff
@@ -211,38 +215,45 @@ class DeviceTrackerPlatform:
     config: dict = attr.ib()
 
     @property
-    def type(self):
+    def type(self) -> str | None:
         """Return platform type."""
-        for methods, platform_type in ((self.LEGACY_SETUP, PLATFORM_TYPE_LEGACY),):
-            for meth in methods:
-                if hasattr(self.platform, meth):
-                    return platform_type
-
+        methods, platform_type = self.LEGACY_SETUP, PLATFORM_TYPE_LEGACY
+        for method in methods:
+            if hasattr(self.platform, method):
+                return platform_type
         return None
 
-    async def async_setup_legacy(self, hass, tracker, discovery_info=None):
+    async def async_setup_legacy(
+        self,
+        hass: HomeAssistant,
+        tracker: DeviceTracker,
+        discovery_info: dict[str, Any] | None = None,
+    ) -> None:
         """Set up a legacy platform."""
+        assert self.type == PLATFORM_TYPE_LEGACY
         full_name = f"{DOMAIN}.{self.name}"
         LOGGER.info("Setting up %s", full_name)
         with async_start_setup(hass, [full_name]):
             try:
                 scanner = None
                 setup = None
                 if hasattr(self.platform, "async_get_scanner"):
-                    scanner = await self.platform.async_get_scanner(
+                    scanner = await self.platform.async_get_scanner(  # type: ignore[attr-defined]
```

Comment:
@cdce8p do you have any interesting in contributing more to mypy? 🙂
https://github.com/python/mypy/issues/1424